rfcs/0166-nix-formatting.md

```diff
@@ -0,0 +1,1059 @@
+  ```nix
+  foo = [ {
+    # content
+  } ];
+  ```
+- Be a bit less eager about expanding lists and attribute sets (for example by allowing up to n elements on a single line).
+  ```nix
+  [ x y ]
+  ```
+
+### Binders and inherit
+
+Let bindings and attribute sets share the same syntax for their items, which is discussed here together.
+
+#### Binders
+
+**Description**
+
+Binders have the most special cases to accommodate for many common Nixpkgs idioms.
+Generally, the following styles exist, which are used depending on the kind and size of the value:
+
+```nix
+#1 single line
+foo = "bar";
+
+#2 single line, on a new line
+very.long.foo =
+  function arg1 arg2 arg3;
+
+#3 multi line, starting on the same line
+foo = function {
+  # args
+};
+
+#4 multi line, starting on a new line
+foo =
+  function
+    arg1
+    arg2
+    arg3
+;
+
+#5 simple function arguments (<=2) start on the same line
+add = x: y: {
+  result = x + y;
+};
+
+#6 attribute bindings always start on a new line
+# If we add more attributes, the indentation stays the same, see #7
+split =
+  { x, ... }:
+  {
+    inc = x + 1;
+    dec = x - 1;
+  };
+
+#7 with enough attribute bindings, they will start on a new line and become multline 
+# it is common to expand input arguments in a way that does not change the number of function applications necessary
+outputs =
+  {
+    x,
+    y,
+    z,
+    ...
+  }:
+  {
+    result = x + y;
+  };
+```
+
+Notable special cases are:
+
+- Single line values that would not benefit from style #2 keep using #1, even if this makes it go above the line limit. This mostly applies to simple strings and paths.
+- Attribute set values are *always* expanded. This has the consequence of always forcing nested attribute sets to be multiline (even if they would be single line otherwise because they only contain a single item), which usually is desired.
+  ```nix
+  {
+    foo.bar.baz = "qux";
+    foo' = {
+      bar.baz = "qux";
+    };
+  }
+  ```
+- "statement-like" expressions like "let", "if" and "assert" always use #4 (or #1).
+- If the value is a `with` followed by a function application or list or attribute set, try to use #3.
+  ```nix
+  buildInputs = with pkgs; [
+    some
+    dependencies
+  ];
+  ```
+
+**Alternatives**
+
+One could eliminate style #2 by having #4 always start on the first line. This would even reduce indentation in some cases. However, this may look really weird in other cases, especially when the binder is very long:
+
+```nix
+some.very.long.attr = callFunction
+  arg1
+  arg2
+  arg3;
+```
+
+#### inherit
+
+**Description**
+
+The items are either all on the same line, or all on a new line each (with indentation).
+
+**Examples**
+
+```nix
+inherit foo bar baz;
+inherit
+  foo'
+  bar'
+  baz'
+;
+```
+
+#### inherit from
+
+**Description**
+
+- If the inherit target is single-line, it is placed on the same line as the `inherit`, even if the following items do not fit onto one line.
```

Comment:
```suggestion
- If the inherit source is single-line, it is placed on the same line as the `inherit`, even if the following items do not fit onto one line.
```

Not sure if that already has a term, but I'd rather call it a *source*, since that's where we inherit *from*.